client/src/app/gateways/vote-decrypt-gateway.service.ts

```diff
@@ -0,0 +1,359 @@
+        }
+
+        const { ephPublKey, sharedSecret } = await this.generateKeysAndSecret(cryptKey);
+        const hkdfKey = await this.generateHkdfKey(sharedSecret);
+
+        const nonce = crypto.getRandomValues(new Uint8Array(this._nonceSize));
+
+        const encrypted = new Uint8Array(
+            await (crypto.subtle.encrypt({ name: `AES-GCM`, iv: nonce }, hkdfKey, plaintext) as Promise<ArrayBuffer>)
+        );
+        console.log(`encrypted!`);
+
+        return this.createPayload(ephPublKey, nonce, encrypted);
+    }
+
+    public async reVerifyPoll(poll: ViewPoll): Promise<void> {
+        const verificationData: VoteVerificationData[] = [
+            { id: poll.id, signature: poll.votes_signature, raws: poll.votes_raw }
+        ];
+        await this.hasPublicMainKey;
+        await this.verify(verificationData);
+    }
+
+    private async updateVerification(verificationData: VoteVerificationData[]): Promise<void> {
+        const prev = this._lastVoteVerificationData.sort((a, b) => a.id - b.id);
+        const curr = verificationData.sort((a, b) => a.id - b.id);
+
+        const toUnverify: VoteVerificationData[] = [];
+        const toVerify: VoteVerificationData[] = [];
+        let x = 0;
+        for (let y = 0; y < curr.length; y++) {
+            while (prev[x] && prev[x].id < curr[y].id) {
+                toUnverify.push(prev[x]);
+                x++;
+            }
+            if (curr[y].id !== prev[x]?.id || JSON.stringify(curr[y]) !== JSON.stringify(prev[x])) {
+                toVerify.push(curr[y]);
+            }
+            if (curr[y].id === prev[x]?.id) {
+                x++;
+            }
+        }
+        for (let z = x; z < prev.length; z++) {
+            toUnverify.push(prev[z]);
+        }
+
+        console.log(`${toVerify.length ? `Verifying ${toVerify.length} polls...` : `No polls to verify.`}`);
+        this.verify(toVerify);
+        this.unverify(toUnverify);
+    }
+
+    private async verify(toVerify: VoteVerificationData[]): Promise<void> {
+        const couldntVerify: number[] = [];
+        for (let date of toVerify) {
+            const model = this.repo.getViewModel(date.id);
+            if (!model) {
+                couldntVerify.push(date.id);
+                continue;
+            }
+            const signature = uInt8Enc(date.signature);
+            const raws = uInt8Enc(date.raws, false);
+            let resultVerificationErrors: string[] = [];
+            try {
+                const verified = await verify(signature, raws, this._publicMainKey);
+                resultVerificationErrors = verified ? resultVerificationErrors : [`Signature verification failed`];
+            } catch (e) {
+                resultVerificationErrors.push(`Signature verification failed with error: "${e.message}"`);
+            }
+            try {
+                resultVerificationErrors = resultVerificationErrors.concat(this.verifyBackendPollResults(model));
```

Comment:
I think it might be better to handle the throwing method call within `verifyBackendPollResults` as it returns a error array already.

Reply:
It may be more practical to write into the documentation of `verifyBackendPollResult` that the function may throw an error from the `getSortedVoteArrays` function